## 第十周作业

#### 提交人学号：10235501454
#### 提交人姓名：陈子谦

#### 作业题目：
    阅读《概率论与数理统计教程》例4.3.1和例4.3.4，分别用随机投点法和平均值法计算下列两个定积分
    并将其与精确值进行比较。
    解答过程中要求：
    (a) 积分结果保留六位小数；
    (b) 在两种蒙特卡洛方法中，令 n 均为 自己的学号后三位加上 12345；并在解答过程中给出n的取值。例如，学号后三位为001，所使用的n为12346。

In [14]:
import numpy as np
from scipy.integrate import quad

# 定义函数
def f(x):
    return (np.exp(x) - 1) / (np.exp(1) - 1)

def g(x):
    return np.exp(x)

# 随机投点法
def monte_carlo_integration(func, a, b, n):
    points = np.random.uniform(a, b, n)
    values = func(points)
    return (b - a) * np.mean(values)

# 平均值法
def average_value_method(func, a, b, n):
    points = np.linspace(a, b, n)
    values = func(points)
    return (b - a) * np.mean(values)

# 精确积分计算
def exact_integral(func, a, b):
    result, _ = quad(func, a, b)
    return result

n = 454 + 12345

# 计算积分
J1_monte_carlo = monte_carlo_integration(f, 0, 1, n)
J1_average_value = average_value_method(f, 0, 1, n)
J1_exact = exact_integral(f, 0, 1)

J2_monte_carlo = monte_carlo_integration(g, -1, 1, n)
J2_average_value = average_value_method(g, -1, 1, n)
J2_exact = exact_integral(g, -1, 1)

# 打印结果
print(f"J1 (蒙特卡洛方法): {J1_monte_carlo:.6f}")
print(f"J1 (平均值法): {J1_average_value:.6f}")
print(f"J1 (精确值): {J1_exact:.6f}")

print(f"J2 (蒙特卡洛方法): {J2_monte_carlo:.6f}")
print(f"J2 (平均值法): {J2_average_value:.6f}")
print(f"J2 (精确值): {J2_exact:.6f}")

print(f"N = {n}")


J1 (蒙特卡洛方法): 0.418403
J1 (平均值法): 0.418030
J1 (精确值): 0.418023
J2 (蒙特卡洛方法): 2.348817
J2 (平均值法): 2.350460
J2 (精确值): 2.350402
N = 12799


##### 最后我们来验证猜想3：
    代码使用随机投点法和平均值法来估算这两个积分，并与精确值进行比较。
    这是一次代码的实验结果：
        J1 (蒙特卡洛方法): 0.418403
        J1 (平均值法): 0.418030
        J1 (精确值): 0.418023
        J2 (蒙特卡洛方法): 2.348817
        J2 (平均值法): 2.350460
        J2 (精确值): 2.350402
        N = 12799